From version 6.3 to 6.4 long messages to "Gradle ANTLR Worker" stopped work.

Updating a project from Gradle 6.2.2 to 7.2.1, it was noticed that the communication to the "Gradle ANTLR Worker" stopped work.

During the investigation process, it was noticed that the problem occur when the content of the output buffer of KryoBackedEncoder overflows. In the previous version the overflows also occur but without problem.

### Expected Behavior
The ANTLR plugin should work with large number of files, as it works on version 6.3.

### Current Behavior
I tested all versions after the 6.3 and all of them fail with the message below:

```
$ ./gradlew clean generateGrammarSource
> Task :generateGrammarSource FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateGrammarSource'.
> Failed to run Gradle ANTLR Worker
   > No response was received from Gradle ANTLR Worker but the worker process has finished.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 848ms
2 actionable tasks: 1 executed, 1 up-to-date

```

### Context
<!--- Which version did this stop working with? Which version do you know it works? -->
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

### Steps to Reproduce 
https://github.com/makall/gradle-antlr-test
* Sucess: `./gradlew wrapper --gradle-version=6.3 && ./gradlew clean generateGrammarSource`
* Fail: `./gradlew wrapper --gradle-version=6.4 && ./gradlew clean generateGrammarSource`

### Your Environment
Build scan URL:
* Success: https://gradle.com/s/3lnhtlwevpyoe
* Fail: https://gradle.com/s/3lp3paoav2qds

--- a/subprojects/messaging/src/main/java/org/gradle/internal/serialize/kryo/KryoBackedDecoder.java
+++ b/subprojects/messaging/src/main/java/org/gradle/internal/serialize/kryo/KryoBackedDecoder.java
@@ -168,6 +168,8 @@ public class KryoBackedDecoder extends AbstractDecoder implements Decoder, Close
     public <T> T decodeChunked(DecodeAction<Decoder, T> decodeAction) throws EOFException, Exception {
         if (nested == null) {
             nested = new KryoBackedDecoder(new InputStream() {
+                private int leftover = 0;
+
                 @Override
                 public int read() throws IOException {
                     throw new UnsupportedOperationException();
@@ -175,14 +177,21 @@ public class KryoBackedDecoder extends AbstractDecoder implements Decoder, Close
 
                 @Override
                 public int read(byte[] buffer, int offset, int length) throws IOException {
+                    if (leftover > 0) {
+                        int count = Math.min(leftover, length);
+                        leftover -= count;
+                        readBytes(buffer, offset, count);
+                        return count;
+                    }
+
                     int count = readSmallInt();
                     if (count == 0) {
                         // End of stream has been reached
                         return -1;
                     }
                     if (count > length) {
-                        // For now, assume same size buffers used to read and write
-                        throw new UnsupportedOperationException();
+                        leftover = count - length;
+                        count = length;
                     }
                     readBytes(buffer, offset, count);
                     return count;

